shocker.c

```diff
@@ -1,60 +1,161 @@
+  fail_sethostname:
+    exit(EXIT_FAILURE);
+}
+
+static bool intWriter(char* path, int num) {
+  FILE *fp = fopen(path, "w");
+
+  if (fp == NULL) {
+    fprintf(stderr, "fopen %s ", path);
+    perror("failed with: ");
+    goto fail_fopen;
+  }
+
+  if (fprintf(fp, "%d", num) == -1) {
+    fprintf(stderr, "fprintf %s with %d ", path, num);
+    perror("failed with: ");
+    goto fail_fprintf;
+  }
+
+  fclose(fp);
+  return true;
+
+  fail_fprintf:
+    fclose(fp);
+  fail_fopen:
+    return false;
 }
 
 int run(char *argv[]) {
-  // Fork and exec pattern in now the clone and exec pattern
-  int child_pid = clone(child, child_stack+STACK_SIZE, 
+
+  /* Allocate memory to be used for the stack of the child */
+  char *stack = mmap(NULL, STACK_SIZE, PROT_READ | PROT_WRITE, MAP_PRIVATE | MAP_ANONYMOUS | MAP_STACK, -1, 0);
+
+  if (stack == MAP_FAILED) {
+    perror("mmap failed with: ");
+    goto fail_mmap;
+  }
+
+  // Fork and exec pattern is now the clone and exec pattern
+  int child_pid = clone(child, stack + STACK_SIZE,
       CLONE_NEWUTS | // unix timesharing, allows changing of hostname 
       CLONE_NEWPID | // procfs is one of a few ways to pass process info, OSX has no procfs
       CLONE_NEWNS | // first namespace (for mounts). Private copy of its namespace.
       CLONE_NEWUSER | // user related
       SIGCHLD, // signal sent to the parent when the child terminates
       argv + 1 // drop `run` arg
   );
-  // create /sys/fs/cgroup/pids/shocker prior to running
-  FILE *fp = fopen("/sys/fs/cgroup/pids/shocker/cgroup.procs", "w");
-  fprintf(fp, "%d", child_pid); // write child pid to the cgroup.procs procfs
-  fclose(fp);
-  waitpid(child_pid, NULL , 0); // NULL is whether to write status, 0 
-  umount("/root/container/proc");
-  return 0;
+
+  if (child_pid == -1) {
+    perror("clone failed with: ");
+    goto fail_clone;
+  }
+
+  char *cgroup_dir = "/sys/fs/cgroup/pids/shocker";
+
+  // make a new cgroup
+  if (mkdir(cgroup_dir, 0644) == -1) {
+    perror("mkdir failed with: ");
+    goto fail_mkdir;
+  }
+
+  // Restrain procs in this cgroup to a max of 5 processes
+  if (!intWriter("/sys/fs/cgroup/pids/shocker/pids.max", 5)) goto fail_pidsmax;
+
+  // Put the child in the new cgroup
+  if (!intWriter("/sys/fs/cgroup/pids/shocker/cgroup.procs", child_pid)) goto fail_cgroupprocs;
+
+  if (waitpid(child_pid, NULL , 0) == -1) {
+    perror("waitpid failed with ");
+    goto fail_waitpid;
+  }
+
+  exit(EXIT_SUCCESS);
+
+  fail_waitpid:
+  fail_cgroupprocs:
+  fail_pidsmax:
+    rmdir(cgroup_dir);
+  fail_mkdir:
+    kill(child_pid, SIGTERM);
+  fail_clone:
+    munmap(stack, STACK_SIZE);
+  fail_mmap:
+    exit(EXIT_FAILURE);;
 }
 
-// need just stdio, and string for main
 
+// need just stdio, and string for main
 int main(int argc, char *argv[]) {
   if (argc < 3) {
     puts("Not enough arguments.\n");
-    return 1;
+    exit(EXIT_FAILURE);
   }
   argv += 1; // drop binary argument
   if (strcmp(argv[0], "run") == 0) {
```

Comment:
My personal rule is to never ever use the `str` functions and only use the `strn` functions. Now this can seem a bit extreme given cases like these but like 90% of C security problems comes from mishandling of string and bounds. As a result I get in the habit of always using the `strn` functions so that I don't fuck it up at the worst moment. It also forces you think about boundary values for the data you're processing which I generally think is a good thing.